Crash with EMI

### Mod loader

Fabric

### Minecraft version

1.20.1

### Mod version

e16bf9b5065a94bc4edd82b2dd9131e49f2713b9

### Modloader version

e16bf9b5065a94bc4edd82b2dd9131e49f2713b9

### Modpack info

_No response_

### If bug:

- [X] Can you reproduce this issue with relevant mods only?

### If bug: The latest.log file

_No response_

### Issue description

The Minecraft option isn't available when reading mod config. Need to move the TA options to mod config I think
```
Caused by: java.lang.NullPointerException: Cannot invoke "snownee.textanimator.duck.TAOptions.textanimator$getTypewriterMode()" because "net.minecraft.client.Minecraft.getInstance().options" is null
	at snownee.textanimator.TextAnimatorClient.getTypewriterMode(TextAnimatorClient.java:41)
	at snownee.textanimator.util.CommonProxy.iterateFormatted(CommonProxy.java:45)
	at net.minecraft.util.StringDecomposer.handler$zeg000$textanimator$iterateFormatted(StringDecomposer.java:521)
	at net.minecraft.util.StringDecomposer.iterateFormatted(StringDecomposer.java)
	at net.minecraft.client.StringSplitter.splitLines(StringSplitter.java:294)
	at net.minecraft.client.StringSplitter.splitLines(StringSplitter.java:311)
	at dev.emi.emi.config.EmiConfig.getSavedConfig(EmiConfig.java:594)
	at dev.emi.emi.config.EmiConfig.<clinit>(EmiConfig.java:809)
	... 10 more
```

--- a/src/main/java/snownee/textanimator/TextAnimatorClient.java
+++ b/src/main/java/snownee/textanimator/TextAnimatorClient.java
@@ -9,6 +9,12 @@ import net.minecraft.client.OptionInstance;
 import net.minecraft.util.Mth;
 import net.minecraft.world.phys.Vec2;
 import snownee.textanimator.duck.TAOptions;
+import snownee.textanimator.effect.EffectFactory;
+import snownee.textanimator.effect.RainbowEffect;
+import snownee.textanimator.effect.ShakeEffect;
+import snownee.textanimator.effect.WaveEffect;
+import snownee.textanimator.effect.WiggleEffect;
+import snownee.textanimator.typewriter.TypewriterEffect;
 
 public class TextAnimatorClient {
 	public static Vec2[] RANDOM_DIR;
@@ -26,11 +32,16 @@ public class TextAnimatorClient {
 		Collections.shuffle(dirs);
 		RANDOM_DIR = dirs.toArray(Vec2[]::new);
 
-//		registerDisabledScreen(BookEditScreen.class);
+		//		registerDisabledScreen(BookEditScreen.class);
 	}
 
 	public static TextAnimationMode getTextAnimationMode() {
-		OptionInstance<TextAnimationMode> instance = ((TAOptions) Minecraft.getInstance().options).textanimator$getTextAnimation();
+		//noinspection ConstantValue https://github.com/Snownee/TextAnimator/issues/25
+		if (Minecraft.getInstance().options == null) {
+			return TextAnimationMode.ALL;
+		}
+		var instance =
+				((TAOptions) Minecraft.getInstance().options).textanimator$getTextAnimation();
 		if (instance == null) {
 			return TextAnimationMode.ALL;
 		}
@@ -38,7 +49,11 @@ public class TextAnimatorClient {
 	}
 
 	public static TypewriterMode getTypewriterMode() {
-		OptionInstance<TypewriterMode> instance = ((TAOptions) Minecraft.getInstance().options).textanimator$getTypewriterMode();
+		//noinspection ConstantValue https://github.com/Snownee/TextAnimator/issues/25
+		if (Minecraft.getInstance().options == null) {
+			return TypewriterMode.BY_CHAR;
+		}
+		var instance = ((TAOptions) Minecraft.getInstance().options).textanimator$getTypewriterMode();
 		if (instance == null) {
 			return TypewriterMode.BY_CHAR;
 		}
@@ -55,7 +70,7 @@ public class TextAnimatorClient {
 		defaultTypewriterInterval = values[9 - speed];
 	}
 
-//	public static synchronized void registerDisabledScreen(Class<?> screen) {
-//		SCREENS_DISABLED.add(screen);
-//	}
+	//	public static synchronized void registerDisabledScreen(Class<?> screen) {
+	//		SCREENS_DISABLED.add(screen);
+	//	}
 }

--- a/src/main/java/snownee/textanimator/TextAnimatorClient.java
+++ b/src/main/java/snownee/textanimator/TextAnimatorClient.java
@@ -5,16 +5,9 @@ import java.util.Collections;
 import java.util.List;
 
 import net.minecraft.client.Minecraft;
-import net.minecraft.client.OptionInstance;
 import net.minecraft.util.Mth;
 import net.minecraft.world.phys.Vec2;
 import snownee.textanimator.duck.TAOptions;
-import snownee.textanimator.effect.EffectFactory;
-import snownee.textanimator.effect.RainbowEffect;
-import snownee.textanimator.effect.ShakeEffect;
-import snownee.textanimator.effect.WaveEffect;
-import snownee.textanimator.effect.WiggleEffect;
-import snownee.textanimator.typewriter.TypewriterEffect;
 
 public class TextAnimatorClient {
 	public static Vec2[] RANDOM_DIR;

--- a/src/main/java/snownee/textanimator/mixin/StringDecomposerAccess.java
+++ b/src/main/java/snownee/textanimator/mixin/StringDecomposerAccess.java
@@ -1,0 +1,16 @@
+package snownee.textanimator.mixin;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Invoker;
+
+import net.minecraft.network.chat.Style;
+import net.minecraft.util.FormattedCharSink;
+import net.minecraft.util.StringDecomposer;
+
+@Mixin(StringDecomposer.class)
+public interface StringDecomposerAccess {
+	@Invoker
+	static boolean callFeedChar(Style p_14333_, FormattedCharSink p_14334_, int p_14335_, char p_14336_) {
+		throw new IllegalAccessError("Shouldn't be called");
+	}
+}

--- a/src/main/java/snownee/textanimator/mixin/client/FontAccess.java
+++ b/src/main/java/snownee/textanimator/mixin/client/FontAccess.java
@@ -1,0 +1,38 @@
+package snownee.textanimator.mixin.client;
+
+import org.joml.Matrix4f;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Accessor;
+import org.spongepowered.asm.mixin.gen.Invoker;
+
+import com.mojang.blaze3d.vertex.VertexConsumer;
+
+import net.minecraft.client.gui.Font;
+import net.minecraft.client.gui.font.FontSet;
+import net.minecraft.client.gui.font.glyphs.BakedGlyph;
+import net.minecraft.resources.ResourceLocation;
+
+@Mixin(Font.class)
+public interface FontAccess {
+	@Accessor
+	boolean getFilterFishyGlyphs();
+
+	@Invoker
+	FontSet callGetFontSet(ResourceLocation resourceLocation);
+
+	@Invoker
+	void callRenderChar(
+			BakedGlyph bakedGlyph,
+			boolean bl,
+			boolean bl2,
+			float f,
+			float g,
+			float h,
+			Matrix4f matrix4f,
+			VertexConsumer vertexConsumer,
+			float i,
+			float j,
+			float k,
+			float l,
+			int m);
+}

--- a/src/main/java/snownee/textanimator/mixin/client/StringRenderOutputMixin.java
+++ b/src/main/java/snownee/textanimator/mixin/client/StringRenderOutputMixin.java
@@ -74,8 +74,8 @@ public abstract class StringRenderOutputMixin {
 		float b;
 		float g;
 		float r;
-		FontSet fontSet = this$0.getFontSet(style.getFont());
-		GlyphInfo glyphInfo = fontSet.getGlyphInfo(codepoint, this$0.filterFishyGlyphs);
+		FontSet fontSet = ((FontAccess) this$0).callGetFontSet(style.getFont());
+		GlyphInfo glyphInfo = fontSet.getGlyphInfo(codepoint, ((FontAccess) this$0).getFilterFishyGlyphs());
 		BakedGlyph bakedGlyph = style.isObfuscated() && codepoint != 32 ? fontSet.getRandomGlyph(glyphInfo) : fontSet.getGlyph(codepoint);
 		boolean bold = style.isBold();
 		float a = this.a;
@@ -120,7 +120,7 @@ public abstract class StringRenderOutputMixin {
 			a = settings.a;
 			if (a != 0) {
 				VertexConsumer vertexConsumer = this.bufferSource.getBuffer(bakedGlyph.renderType(this.mode));
-				this$0.renderChar(
+				((FontAccess) this$0).callRenderChar(
 						bakedGlyph,
 						bold,
 						style.isItalic(),

--- a/src/main/java/snownee/textanimator/util/CommonProxy.java
+++ b/src/main/java/snownee/textanimator/util/CommonProxy.java
@@ -28,6 +28,7 @@ import snownee.textanimator.TypewriterMode;
 import snownee.textanimator.duck.TAStyle;
 import snownee.textanimator.effect.Effect;
 import snownee.textanimator.effect.params.Params;
+import snownee.textanimator.mixin.StringDecomposerAccess;
 
 @Mod("textanimator")
 public class CommonProxy {
@@ -155,7 +156,7 @@ public class CommonProxy {
 				}
 				return false;
 			}
-			if (StringDecomposer.feedChar(curStyle, formattedCharSink, k, c)) {
+			if (StringDecomposerAccess.callFeedChar(curStyle, formattedCharSink, k, c)) {
 				continue;
 			}
 			return false;

--- a/src/main/resources/META-INF/accesstransformer.cfg
+++ b/src/main/resources/META-INF/accesstransformer.cfg
@@ -1,11 +1,5 @@
 public net.minecraft.client.gui.Font$StringRenderOutput
-public net.minecraft.client.gui.Font f_242994_ # filterFishyGlyphs
-public net.minecraft.client.gui.Font m_92863_(Lnet/minecraft/resources/ResourceLocation;)Lnet/minecraft/client/gui/font/FontSet; # getFontSet
-public net.minecraft.client.gui.Font m_253238_(Lnet/minecraft/client/gui/font/glyphs/BakedGlyph;ZZFFFLorg/joml/Matrix4f;Lcom/mojang/blaze3d/vertex/VertexConsumer;FFFFI)V # renderChar
-
-public net.minecraft.network.chat.Style <init>(Lnet/minecraft/network/chat/TextColor;Ljava/lang/Boolean;Ljava/lang/Boolean;Ljava/lang/Boolean;Ljava/lang/Boolean;Ljava/lang/Boolean;Lnet/minecraft/network/chat/ClickEvent;Lnet/minecraft/network/chat/HoverEvent;Ljava/lang/String;Lnet/minecraft/resources/ResourceLocation;)V # <init>
 
 public net.minecraft.client.Options$FieldAccess
 
-public net.minecraft.util.StringDecomposer m_14332_(Lnet/minecraft/network/chat/Style;Lnet/minecraft/util/FormattedCharSink;IC)Z # feedChar
 public net.minecraft.client.StringSplitter$LineComponent

--- a/src/main/resources/textanimator.mixins.json
+++ b/src/main/resources/textanimator.mixins.json
@@ -5,11 +5,13 @@
 	"compatibilityLevel": "JAVA_16",
 	"refmap": "textanimator.refmap.json",
 	"mixins": [
+		"StringDecomposerAccess",
 		"StringDecomposerMixin",
 		"StyleMixin",
 		"StyleMixin$SerializerMixin"
 	],
 	"client": [
+		"client.FontAccess",
 		"client.FontMixin",
 		"client.LiteralContentsMixin",
 		"client.OptionsMixin",

